pkg/config/remote_config.go

```diff
@@ -0,0 +1,127 @@
+package config
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"github.com/prometheus/common/config"
+)
+
+// supported remote config provider schemes
+const (
+	HTTP  = "http"
+	HTTPS = "https"
+	// TODO: add s3, gcs, blob, and git providers backed by go-fsimpl
+)
+
+// RemoteOpts struct contains agent remote config options
+type RemoteOpts struct {
+	ExpandEnv bool
+
+	URL              *url.URL
+	HTTPClientConfig *config.HTTPClientConfig
+}
+
+// RemoteProvider ...
+type RemoteProvider interface {
+	Retrieve() (*Config, error)
+}
+
+// RemoteConfig ...
+type RemoteConfig struct {
+	RemoteProvider
+}
+
+// NewRemoteConfig ...
+func NewRemoteConfig(rawURL string, opts *RemoteOpts) (*RemoteConfig, error) {
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return nil, err
+	}
+	if opts == nil {
+		// Default provider opts
+		opts = &RemoteOpts{
+			ExpandEnv: false,
+		}
+	}
+	opts.URL = u
+
+	switch u.Scheme {
+	case "":
+		// if no scheme, assume local file path, return nil and let caller handle.
+		return nil, nil
+	case HTTP, HTTPS:
+		return &RemoteConfig{
+			RemoteProvider: newHTTPProvider(opts),
+		}, nil
+	}
+	return nil, fmt.Errorf("remote config scheme not supported: %s", u.Scheme)
+}
+
+// Remote Config Providers
+// httpP - http/https provider
+type httpP struct {
+	myURL            *url.URL
+	expandEnv        bool
+	httpClientConfig *config.HTTPClientConfig
+}
+
+func newHTTPProvider(opts *RemoteOpts) *httpP {
+	return &httpP{
+		myURL:            opts.URL,
+		expandEnv:        opts.ExpandEnv,
+		httpClientConfig: opts.HTTPClientConfig,
+	}
+}
+
+// Retrieve implements RemoteProvider and fetches the config
+// TODO: token auth, oauth2, etc.
+func (p httpP) Retrieve() (*Config, error) {
+	var (
+		bb  []byte
+		err error
+
+		request  *http.Request
+		response *http.Response
+		client   *http.Client
+
+		result = &Config{}
+	)
```

Comment:
very minor nit: I don't think you need to declare every variable you might use at the top like this, but I also don't really care that much

Reply:
Typically just got into a habit of doing this since it can be helpful for GC, especially when re-assigning to declared variables in a loop that gets executed lots of times but in this case probably isn't needed.